File: python/src/refreshing_cache.py
import math
import random
import threading
import time
from concurrent.futures import Future, TimeoutError
from dataclasses import dataclass


@dataclass(repr=False)
class CacheEntry:
    value: object
    expires_at: float
    refresh_at: float


def _schedule(delay, callback):
    timer = threading.Timer(delay, callback)
    timer.daemon = True
    timer.start()
    return timer


class RefreshingCache:
    def __init__(self, load, *, clock=time.time, schedule=_schedule, jitter=random.random,
                 on_failure=lambda: None, wait_timeout=2.5):
        self._load = load
        self._clock = clock
        self._schedule = schedule
        self._jitter = jitter
        self._on_failure = on_failure
        self._wait_timeout = wait_timeout
        self._lock = threading.RLock()
        self._entry = None
        self._inflight = None
        self._timer = None
        self._retry_at = 0
        self._failures = 0
        self._closed = False

    def get(self):
        with self._lock:
            if self._closed:
                raise ValueError("provider credential unavailable")
            now = self._clock()
            if self._entry and self._entry.expires_at > now:
                if self._entry.refresh_at <= now and self._retry_at <= now:
                    self._begin_refresh()
                return self._entry.value
            if self._inflight is None and self._retry_at > now:
                raise ValueError("provider credential unavailable")
            future = self._begin_refresh()
        try:
            return future.result(timeout=self._wait_timeout)
        except TimeoutError:
            # A waiter does not cancel the shared refresh needed by other requests.
            raise ValueError("provider credential unavailable") from None

    def refresh(self):
        with self._lock:
            if self._closed or self._retry_at > self._clock():
                raise ValueError("provider credential unavailable")
            return self._begin_refresh()

    def _begin_refresh(self):
        if self._inflight is not None:
            return self._inflight
        if self._timer is not None:
            self._timer.cancel()
            self._timer = None
        future = Future()
        self._inflight = future
        threading.Thread(target=self._run_refresh, args=(future,), daemon=True).start()
        return future

    def _run_refresh(self, future):
        entry = None
        failed = False
        try:
            entry = self._load()
            if (not isinstance(entry, CacheEntry) or not math.isfinite(entry.expires_at)
                    or entry.expires_at <= self._clock() or not math.isfinite(entry.refresh_at)):
                raise ValueError("provider credential unavailable")
        except Exception:
            failed = True
        with self._lock:
            if self._closed:
                if not future.done():
                    future.set_exception(ValueError("provider credential unavailable"))
                self._inflight = None
                return
            if failed:
                self._failures += 1
                backoff = min(60, 5 * 2 ** min(self._failures - 1, 4))
                self._retry_at = self._clock() + backoff * (1 + self._jitter() * 0.2)
                self._on_failure()
            else:
                self._entry = entry
                self._failures = 0
                self._retry_at = 0
            self._inflight = None
            next_refresh = self._retry_at if failed else entry.refresh_at
            self._timer = self._schedule(max(1, next_refresh - self._clock()), self._scheduled_refresh)
            if failed:
                future.set_exception(ValueError("provider credential unavailable"))
            else:
                future.set_result(entry.value)

    def _scheduled_refresh(self):
        with self._lock:
            self._timer = None
            if not self._closed:
                self._begin_refresh()

    def close(self):
        with self._lock:
            self._closed = True
            if self._timer is not None:
                self._timer.cancel()
                self._timer = None
            self._entry = None


def token_entry(token, now):
    value = getattr(token, "token", None)
    expiry = getattr(token, "expires_on", None)
    if (not isinstance(value, str) or not value.strip() or type(expiry) not in (int, float)
            or not math.isfinite(expiry) or expiry <= now + 30):
        raise ValueError("provider credential unavailable")
    expires_at = expiry - 30
    refresh_at = expiry - 300
    hint = getattr(token, "refresh_on", None)
    if type(hint) in (int, float) and math.isfinite(hint):
        refresh_at = min(refresh_at, hint)
    if refresh_at <= now:
        refresh_at = now + max(1, min(60, (expires_at - now) / 2))
    return CacheEntry(token, expires_at, refresh_at)

File: dotnet/Src/RefreshingCache.cs
namespace Epp.Otp;

internal static class CredentialCachePolicy
{
    internal static readonly TimeSpan AcquisitionTimeout = TimeSpan.FromSeconds(2.5);
    internal static readonly TimeSpan SecretTtl = TimeSpan.FromMinutes(5);
    internal static readonly TimeSpan SecretRefreshInterval = TimeSpan.FromMinutes(4);
    internal static readonly TimeSpan TokenExpirySkew = TimeSpan.FromSeconds(30);
    internal static readonly TimeSpan TokenRefreshLead = TimeSpan.FromMinutes(5);
    internal static readonly TimeSpan MinRefreshDelay = TimeSpan.FromSeconds(1);
    internal static readonly TimeSpan MaxRefreshDelay = TimeSpan.FromSeconds(60);
    internal static readonly TimeSpan InitialRetryDelay = TimeSpan.FromSeconds(5);
    internal static readonly TimeSpan MaxRetryDelay = TimeSpan.FromSeconds(60);
    internal static readonly TimeSpan MaxTimerDelay = TimeSpan.FromMilliseconds(int.MaxValue);
    internal const int MaxRetryExponent = 4;
    internal const double RetryJitterRatio = 0.2;
}

internal sealed record CredentialCacheEntry<T>(T Value, DateTimeOffset ExpiresAt, DateTimeOffset RefreshAt)
{
    public override string ToString() => nameof(CredentialCacheEntry<T>);
}

internal sealed class RefreshingCache<T> : IDisposable
{
    private readonly object _gate = new();
    private readonly Func<CancellationToken, Task<CredentialCacheEntry<T>>> _load;
    private readonly TimeProvider _clock;
    private readonly Func<double> _random;
    private readonly Action _onFailure;
    private CredentialCacheEntry<T>? _entry;
    private Task<T>? _inFlight;
    private CancellationTokenSource? _acquisition;
    private ITimer? _timer;
    private DateTimeOffset _retryAt;
    private int _failures;
    private bool _closed;

    internal RefreshingCache(Func<CancellationToken, Task<CredentialCacheEntry<T>>> load,
        Action onFailure, TimeProvider? clock = null, Func<double>? random = null)
    {
        _load = load;
        _onFailure = onFailure;
        _clock = clock ?? TimeProvider.System;
        _random = random ?? Random.Shared.NextDouble;
    }

    private static Exception Unavailable() => new InvalidOperationException("provider credential unavailable");

    internal Task<T> GetAsync(CancellationToken cancellationToken = default)
    {
        Task<T> pending;
        lock (_gate)
        {
            if (_closed) return Task.FromException<T>(Unavailable());
            var now = _clock.GetUtcNow();
            if (_entry is not null && _entry.ExpiresAt > now)
            {
                if (_entry.RefreshAt <= now && _retryAt <= now) _ = ObserveAsync(StartRefresh());
                return Task.FromResult(_entry.Value);
            }
            if (_inFlight is null && _retryAt > now) return Task.FromException<T>(Unavailable());
            pending = StartRefresh();
        }
        return cancellationToken.CanBeCanceled ? pending.WaitAsync(cancellationToken) : pending;
    }

    internal Task<T> RefreshAsync()
    {
        lock (_gate)
        {
            if (_closed || _retryAt > _clock.GetUtcNow()) return Task.FromException<T>(Unavailable());
            return StartRefresh();
        }
    }

    private Task<T> StartRefresh()
    {
        if (_inFlight is not null) return _inFlight;
        _timer?.Dispose();
        _timer = null;
        var completion = new TaskCompletionSource<T>(TaskCreationOptions.RunContinuationsAsynchronously);
        _inFlight = completion.Task;
        _acquisition = new CancellationTokenSource(CredentialCachePolicy.AcquisitionTimeout, _clock);
        _ = RunRefreshAsync(completion, _acquisition);
        return completion.Task;
    }

    private async Task RunRefreshAsync(TaskCompletionSource<T> completion, CancellationTokenSource cancellation)
    {
        CredentialCacheEntry<T>? entry = null;
        try
        {
            entry = await _load(cancellation.Token).WaitAsync(cancellation.Token).ConfigureAwait(false);
            if (entry.ExpiresAt <= _clock.GetUtcNow()) entry = null;
        }
        catch (Exception)
        {
            // Emit only the fixed failure below; SDK exceptions may contain credentials.
        }
        lock (_gate)
        {
            _inFlight = null;
            _acquisition = null;
            if (_closed || cancellation.IsCancellationRequested) entry = null;
            cancellation.Dispose();
            if (!_closed)
            {
                if (entry is null)
                {
                    _failures++;
                    var exponent = Math.Min(_failures - 1, CredentialCachePolicy.MaxRetryExponent);
                    var backoff = Math.Min(CredentialCachePolicy.MaxRetryDelay.TotalSeconds,
                        CredentialCachePolicy.InitialRetryDelay.TotalSeconds * Math.Pow(2, exponent));
                    _retryAt = _clock.GetUtcNow().AddSeconds(backoff * (1 + _random() * CredentialCachePolicy.RetryJitterRatio));
                    _onFailure();
                }
                else
                {
                    _entry = entry;
                    _failures = 0;
                    _retryAt = default;
                }
                var next = entry is null ? _retryAt : entry.RefreshAt;
                var wait = Math.Clamp((next - _clock.GetUtcNow()).TotalMilliseconds,
                    CredentialCachePolicy.MinRefreshDelay.TotalMilliseconds, CredentialCachePolicy.MaxTimerDelay.TotalMilliseconds);
                _timer = _clock.CreateTimer(_ => ScheduledRefresh(), null, TimeSpan.FromMilliseconds(wait), Timeout.InfiniteTimeSpan);
            }
            if (entry is null) completion.TrySetException(Unavailable());
            else completion.TrySetResult(entry.Value);
        }
    }

    private void ScheduledRefresh()
    {
        lock (_gate)
        {
            _timer?.Dispose();
            _timer = null;
            if (!_closed) _ = ObserveAsync(StartRefresh());
        }
    }

    private static async Task ObserveAsync(Task<T> task)
    {
        // Refresh failures are already reported; a timer has no request awaiting the result.
        try { await task.ConfigureAwait(false); }
        catch (Exception) { }
    }

    public void Dispose()
    {
        lock (_gate)
        {
            _closed = true;
            _timer?.Dispose();
            _timer = null;
            _acquisition?.Cancel();
            _entry = null;
        }
    }
}
